datahub-web-react/src/app/ingest/source/builder/RecipeForm/RecipeForm.tsx

```diff
@@ -0,0 +1,139 @@
+import { Button, Collapse, Form, message, Typography } from 'antd';
+import React from 'react';
+import YAML from 'yamljs';
+import { ApiOutlined, FilterOutlined, SettingOutlined } from '@ant-design/icons';
+import styled from 'styled-components/macro';
+import { jsonToYaml } from '../../utils';
+import { RecipeField, RECIPE_FIELDS } from './constants';
+import FormField from './FormField';
+
+export const ControlsContainer = styled.div`
+    display: flex;
+    justify-content: space-between;
+    margin-top: 12px;
+`;
+
+const StyledCollapse = styled(Collapse)`
+    margin-bottom: 16px;
+
+    .ant-collapse-header {
+        font-size: 14px;
+        font-weight: bold;
+        padding: 12px 0;
+    }
+`;
+
+const HeaderTitle = styled.span`
+    margin-left: 8px;
+`;
+
+const MarginWrapper = styled.div`
+    margin-left: 20px;
+`;
+
+function getInitialValues(displayRecipe: string, allFields: any[]) {
+    const initialValues = {};
+    let recipeObj;
+    try {
+        recipeObj = YAML.parse(displayRecipe);
+    } catch (e) {
+        message.warn('Found invalid YAML. Please check your recipe configuration.');
+        return {};
+    }
+    if (recipeObj) {
+        allFields.forEach((field) => {
+            initialValues[field.name] = field.getValueFromRecipe(recipeObj);
+        });
+    }
+
+    return initialValues;
+}
+
+function SectionHeader({ icon, text }: { icon: any; text: string }) {
+    return (
+        <span>
+            {icon}
+            <HeaderTitle>{text}</HeaderTitle>
+        </span>
+    );
+}
+
+function shouldRenderFilterSectionHeader(field: RecipeField, index: number, filterFields: RecipeField[]) {
+    if (index === 0 && field.section) return true;
+    if (field.section && filterFields[index - 1].section !== field.section) return true;
+    return false;
+}
+
+interface Props {
+    type: string;
+    isEditing: boolean;
+    displayRecipe: string;
+    setStagedRecipe: (recipe: string) => void;
+    onClickNext: () => void;
+    goToPrevious?: () => void;
+}
+
+function RecipeForm(props: Props) {
+    const { type, isEditing, displayRecipe, setStagedRecipe, onClickNext, goToPrevious } = props;
+    const { fields, advancedFields, filterFields } = RECIPE_FIELDS[type];
+    const allFields = [...fields, ...advancedFields, ...filterFields];
+
+    function updateFormValues(_changedValues: any, allValues: any) {
+        let updatedValues = YAML.parse(displayRecipe);
+        allFields.forEach((field) => {
+            updatedValues = field.setValueOnRecipe(updatedValues, allValues[field.name]);
+        });
```

Comment:
couldn't we just iterate over `_changedValues`? would that be faster?

Reply:
why yes it would